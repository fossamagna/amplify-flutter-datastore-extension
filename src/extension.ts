import path from "node:path";
import type { DocumentNode } from "graphql";
import { codegen } from "@graphql-codegen/core";
import type { Types } from "@graphql-codegen/plugin-helpers";
import { DART_SCALAR_MAP } from "@aws-amplify/appsync-modelgen-plugin/lib/scalars";
import { plugin, addToSchema } from "./plugin";

export type GenerateExtensionsOptions = {
  schema: DocumentNode;
  directives: string;
  baseOutputDir: string;
  // feature flags
  generateIndexRules?: boolean;
  emitAuthProvider?: boolean;
  useExperimentalPipelinedTranformer?: boolean;
  transformerVersion?: number;
  respectPrimaryKeyAttributesOnConnectionField?: boolean;
  generateModelsForLazyLoadAndCustomSelectionSet?: boolean;
  addTimestampFields?: boolean;
  handleListNullabilityTransparently?: boolean;
};

export type GeneratedOutput = {
  [filepath: string]: string;
};

export async function generateExtensions(
  options: GenerateExtensionsOptions,
): Promise<GeneratedOutput> {
  const {
    schema,
    directives,
    // feature flags
    generateIndexRules = true,
    emitAuthProvider = true,
    useExperimentalPipelinedTranformer = true,
    transformerVersion = 2,
    respectPrimaryKeyAttributesOnConnectionField = true,
    generateModelsForLazyLoadAndCustomSelectionSet = true,
    addTimestampFields = true,
    handleListNullabilityTransparently = true,
  } = options;

  const codegenOptions: Types.GenerateOptions = {
    filename: path.join(options.baseOutputDir, "DataStoreExtension.dart"),
    plugins: [
      {
        appSyncDartExtensionPlugin: {},
      },
    ],
    schema,
    documents: [],
    config: {
      directives,
      scalars: { ...DART_SCALAR_MAP },
      target: "dart",
      isTimestampFieldsAdded: addTimestampFields,
      emitAuthProvider,
      generateIndexRules,
      handleListNullabilityTransparently,
      usePipelinedTransformer: useExperimentalPipelinedTranformer,
      transformerVersion,
      respectPrimaryKeyAttributesOnConnectionField,
      generateModelsForLazyLoadAndCustomSelectionSet,
    },
    pluginMap: {
      appSyncDartExtensionPlugin: {
        plugin,
        addToSchema,
      },
    },
  };
  const generatedCode = await codegen(codegenOptions);
  return { [codegenOptions.filename]: generatedCode };
}
