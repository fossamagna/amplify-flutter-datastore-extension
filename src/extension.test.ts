import { parse } from "graphql";
import { generateExtensions, GenerateExtensionsOptions } from "./extension";

describe("extension", () => {
  it("should be generate extensions", async () => {
    const typeDefs = /* GraphQL */ `
      type Todo @model {
        id: ID!
        name: String!
        description: String
      }
    `;
    const schema = parse(typeDefs);
    const options: GenerateExtensionsOptions = {
      baseOutputDir: "lib/models",
      directives: "",
      schema,
    };
    const output = await generateExtensions(options);
    expect(output).toMatchSnapshot();
  });

  it("should be generate extensions with primaryKey and sortKey", async () => {
    const typeDefs = /* GraphQL */ `
      type Inventory @model {
        productID: ID! @primaryKey(sortKeyFields: ["warehouseID"])
        warehouseID: ID!
        InventoryAmount: Int!
      }
    `;
    const schema = parse(typeDefs);
    const options: GenerateExtensionsOptions = {
      baseOutputDir: "lib/models",
      directives: "",
      schema,
    };
    const output = await generateExtensions(options);
    expect(output).toMatchSnapshot();
  });

  it("should not be generate extensions with primaryKey and sortKey when respectPrimaryKeyAttributesOnConnectionField is false", async () => {
    const typeDefs = /* GraphQL */ `
      type Inventory @model {
        productID: ID! @primaryKey(sortKeyFields: ["warehouseID"])
        warehouseID: ID!
        InventoryAmount: Int!
      }
    `;
    const schema = parse(typeDefs);
    const options: GenerateExtensionsOptions = {
      baseOutputDir: "lib/models",
      directives: "",
      schema,
      respectPrimaryKeyAttributesOnConnectionField: false,
    };
    const output = await generateExtensions(options);
    expect(output).toMatchSnapshot();
  });
});
