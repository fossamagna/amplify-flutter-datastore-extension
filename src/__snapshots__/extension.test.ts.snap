// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extension should be generate extensions 1`] = `
{
  "lib/models/DataStoreExtension.dart": "// NOTE: This file is generated and may not follow lint rules defined in your app
// Generated files can be excluded from analysis in analysis_options.yaml
// For more info, see: https://dart.dev/guides/language/analysis-options#excluding-code-from-analysis

// ignore_for_file: public_member_api_docs, annotate_overrides, dead_code, dead_codepublic_member_api_docs, depend_on_referenced_packages, file_names, library_private_types_in_public_api, no_leading_underscores_for_library_prefixes, no_leading_underscores_for_local_identifiers, non_constant_identifier_names, null_check_on_nullable_type_parameter, prefer_adjacent_string_concatenation, prefer_const_constructors, prefer_if_null_operators, prefer_interpolation_to_compose_strings, slash_for_doc_comments, sort_child_properties_last, unnecessary_const, unnecessary_constructor_name, unnecessary_late, unnecessary_new, unnecessary_null_aware_assignments, unnecessary_nullable_for_final_variable_declarations, unnecessary_string_interpolations, use_build_context_synchronously

import 'ModelProvider.dart';
import 'package:amplify_core/amplify_core.dart' as amplify_core;
import 'package:amplify_datastore/amplify_datastore.dart';


/** This is an auto generated extension representing the Todo type in your schema. */
extension TodoExtension on AmplifyDataStore {
  Future<Todo> getTodo(String id) {
    return query(
      Todo.classType,
      where: Todo.MODEL_IDENTIFIER.eq(TodoModelIdentifier(id: id)),
    )
    .then((list) => list.single);
  }
  
  Future<Todo?> getTodoOrNull(String id) {
    return query(
      Todo.classType,
      where: Todo.MODEL_IDENTIFIER.eq(TodoModelIdentifier(id: id)),
    )
    .then((list) => list.singleOrNull);
  }
}",
}
`;
